src/util/schnorr.rs

```diff
@@ -85,4 +89,75 @@ impl TweakedPublicKey {
         &self.0
     }
 
-}
\ No newline at end of file
+}
+
+/// A BIP340-341 serialized schnorr signature with the corresponding hash type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SchnorrSig {
+    /// The underlying schnorr signature
+    pub sig: secp256k1::schnorrsig::Signature,
+    /// The corresponding hash type
+    pub hash_ty: SchnorrSigHashType,
+}
+
+impl SchnorrSig {
+
+    /// Deserialize from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, SchnorrSigError> {
+        if sl.len() == SCHNORRSIG_SIGNATURE_SIZE {
+            // default type
+            let sig = secp256k1::schnorrsig::Signature::from_slice(sl)?;
+            return Ok( SchnorrSig { sig, hash_ty : SchnorrSigHashType::Default });
+        }
+        let (hash_ty, sig) = sl.split_last()
+            .ok_or_else(|| secp256k1::Error::InvalidSignature)?;
+        let hash_ty = SchnorrSigHashType::from_u8(*hash_ty)
+            .map_err(|_| SchnorrSigError::InvalidSighashType(*hash_ty))?;
+        let sig = secp256k1::schnorrsig::Signature::from_slice(sig)?;
+        Ok(SchnorrSig { sig, hash_ty })
+    }
+
+    /// Serialize SchnorrSig
+    pub fn to_vec(&self) -> Vec<u8> {
+        // TODO: add support to serialize to a writer to SerializedSig
+        let mut ser_sig = self.sig.as_ref().to_vec();
+        if let SchnorrSigHashType::Default = self.hash_ty {
+            // default sighash type, don't add extra sighash byte
+        } else {
+            ser_sig.push(self.hash_ty as u8);
+        }
```

Comment:
Overall, this can be simplified into
```suggestion
        if self.hash_ty != SchnorrSigHashType::Default {
           ser_sig.push(self.hash_ty as u8);
        }
```

Reply:
Agreed with all the above nits, but I think the comment and explicit `if/else` have some value here in the form of documentation. I will fix the above nits, but I would like to keep this one as is.